src/data_generators/m3_data_design.py

```diff
@@ -157,19 +166,200 @@
             "order": OrderOptions.random,
             "correlated_sets": CorrelatedSetsOptions.one,
             "shared_prop": 0.75,
-            "random_generator": 8,
+            "random_generator": np.random.default_rng(RNG.integers(1e9)),
         },
     ),
+]
+
+
+class CopulaCorrelationMatrixGenerator:
+    @staticmethod
+    def homogeneous(p: int, rho: float) -> np.ndarray:
+        """Generate a homogeneous correlation matrix.
+
+        Args:
+            p:  Number of pubs.
+            rho:  Homogenous correlation.
+
+        Returns:
+            A p * p matrix with diagonals being 1 and all off-diagonals being rho.
+        """
+        return splinalg.toeplitz([1] + [rho] * (p - 1))
+
+    @staticmethod
+    def autoregressive(p: int, rho: float) -> np.ndarray:
+        """Generate a autoregressive correlation matrix.
+
+        Args:
+            p:  Number of pubs.
+            rho:  Homogenous correlation.
+
+        Returns:
+            A <p * p> matrix C where C[i, j] = rho^|i - j| for any i, j.
+        """
+        return splinalg.toeplitz(np.power(np.array([rho] * p), np.arange(p)))
+
+    @staticmethod
+    def random(
+        p: int, rng: np.random.Generator = np.random.default_rng(0)
+    ) -> np.ndarray:
+        """Randomly, uniformly draw a correlation matrix.
+
+        Given the dimension, all the possible correlation matrices form a space,
+        and there is a geometric measure on this space.  The geoemetric measure
```

Comment:
Could you further explain what you mean by geometric measure?

Reply:
A p * p correlation matrix is in the (p * p) dimensional Euclidean space.  So, the geometric measure of (p * p) dimensional Euclidean space induces a measure on the manifold of p * p correlation matrices.  
This was my interpretation from the abstract + introduction of the paper, and I regret not having checked the math details.  **Now I doubt if the generated matrices are indeed uniform in this sense.**

@chenweiw Could you please chime in?  Per discussion with Chenwei, the paper is not very easy to read.  Let's spend more time reading it, OR do some experiments to confirm the definition of uniformity. WDYT?